### ISSUE TITLE
Fix eslint errors in all exercises

### ISSUE THREAD

**Issue by tejasbubane** [exercism/javascript#480]
I have configured eslint in https://github.com/exercism/javascript/pull/479

But have not fixed errors in any exercise apart from `hello-world`. Instead I added all exercises to [.eslintignore](https://github.com/exercism/javascript/blob/master/.eslintignore). 

The plan is to do this:
1. Pick one exercise (You can pick many but we prefer to have separate PRs so that it is easy to review - but if there are not many changes - you can pick 2-3 at once).
2. Remove the exercise name from `.eslintignore` file
3. Run `npm run lint` - This should show errors in those exercises.
4. Fix those errors - commit and raise a PR 🎉 

Note: The `eslint` configuration is not final. I did not have a good look at all errors, just put a sensible default in place. While fixing, if you encounter some error that you feel we should allow(ignore), feel free to comment on this issue.

**Comment by gargrave** [comment]
I'll help out with some of these over the weekend! 👍 Given how small the overall codebase is, it really shouldn't be too big of a deal to get all of these taken of pretty quickly.

**Comment by gargrave** [comment]
@tarunvelli and I have a handful of PRs open for this--do you need review help? I am happy to help review PRs if that will help expedite things.

**Comment by tejasbubane** [comment]
Thanks a lot for the PRs! I am on a vacation for a week and would be able to look at them next week only. Meanwhile some other maintainers might have a chance to review.

**Comment by matthewmorgan** [comment]
@gargrave @tejasbubane @tarunvelli I'm looking at the PRs now.  Thanks for the patience.

**Comment by ganderzz** [comment]
Trying my hand at helping out if there are no issues with it. :) Should be able to bust out a few.

**Comment by tarunvelli** [comment]
I think the hacktoberfest label should be added to this issue, as it might help get this issue resolved faster

**Comment by Futuro212** [comment]
Hi everyone,

So I'm new with this project and when I try to run the lint command, it shows errors about `Expected linebreaks to be 'LF' but found 'CRLF'`. I'm running this on windows. Should I change some configuration in the eslint?

**Comment by matthewmorgan** [comment]
@Futuro212 let's not comment on closed issues please.  Would you mind opening a separate, new issue for this?  Hopefully someone can help.  Thanks!

**Comment by serixscorpio** [comment]
@matthewmorgan is seems the lint errors on various exercises are not yet all fixed, perhaps re-open this issue? Or do you prefer 1 issue mapping to 1 PR? My PR #505 fixes just 1 exercise.

**Comment by tejasbubane** [comment]
@adamxtokyo Thanks for contributing! We do not want to create separate issues for each PRs. Because the steps outlined in this issue description will be same for all exercises. For more clarity you are advised to look into linked PRs for other exercises. Discussion related to specific exercise as required can be taken up on the PR itself. Otherwise before starting if you have any doubts, you can always comment here.

Regarding contributing guidelines, that is a good point that you brought up. We have quite a few things going on in the CI and elsewhere that we should have a separate document explaining them. I will create a new issue for this thanks! Also I will shortly look into your PR.

**Comment by adamxtokyo** [comment]
Thanks @tejasbubane ! 😄

**Comment by adamxtokyo** [comment]
So, going through the code and linting, I've noticed a few linting rules I think we should change:

**Bitwise operators are not allowed**
Imho this is limiting for more advanced coders. I think we should allow it, or at least downgrade it from an error to a warning (if that makes it pass CI).

**Functions need to be declared before they're used**
Function declarations are hoisted, so this is valid JavaScript, and it makes sense to have the main logic at the top if you're not working on a bigger project and splitting it into modules. Since the exercises are small, I think we should allow it. Maybe this is opinionated though? You're free to disagree with me  😛

**Iterators/generators are not allowed**
The error I'm getting is:
> iterators/generators require regenerator-runtime, which is too heavyweight for this guide to allow them. Separately, loops should be avoided in favor of array iterations  no-restricted-syntax
Not sure what to make of this tbh, maybe they can't be allowed? But I think it's strange to not allow these if it's at all possible. What does everybody else think?

**Comment by tejasbubane** [comment]
> **Bitwise operators are not allowed**
> Imho this is limiting for more advanced coders. I think we should allow it, or at least downgrade it from an error to a warning (if that makes it pass CI).

[According to eslint](https://eslint.org/docs/rules/no-bitwise) the reason this is because people tend to make typos of `|` when they want `||` and same for `&` and `&&`. I think since we have test-cases, we need not worry about this and limit ourselves from using bitwise operators when needed - sometimes they lead to really elegant and efficient solutions. So I agree upon disabling this rule at project level.

**Comment by lukejpreston** [comment]
Are there even any errors left? Also, I could see it here but `eslint --fix` will auto fix loads

**Comment by tejasbubane** [comment]
@lukejpreston 

> Are there even any errors left?

We have disabled eslint errors using [.eslintignore](https://github.com/exercism/javascript/blob/6d2e197d98cd1cbf8c8f963955b843f18a341c5b/.eslintignore) for all exercises and plan to fix errors in 1-2 exercises at a time. This is to make CI happy for now while using it to fix errors by removing the exercises we fix. (See PRs attached to this issue for reference).


> but `eslint --fix` will auto fix loads

It fixes most of them but not all are fixable. You are free to use it but we would like to have each PR fix only 1-2 exercises entirely. This way it is easy to manage/review PRs.

**Comment by lukejpreston** [comment]
thanks, didn't realise the `.eslintignore` file existed

**Comment by thanhcng** [comment]
I'd like to work on this pls
